### PR TITLE
[soci] Fix dependencies in, intern, out

### LINKED ISSUE #34168: Soci build failure
### Operating system

Ubuntu 22.04 (inside docker container)

### Compiler

g++

### Steps to reproduce the behavior

```Shell
https://stackoverflow.com/questions/77140854/how-to-install-soci-db-library-with-vcpkg-on-linux
```

### Failure logs

```
cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake
Soci found:  and 
-- Configuring done
-- Generating done
-- Build files have been written to: /home
root@agh-legion:/home# make
[ 50%] Building CXX object CMakeFiles/Learning_VCPKG.dir/main.cpp.o
[100%] Linking CXX executable Learning_VCPKG
/usr/bin/ld: vcpkg/installed/x64-linux/debug/lib/libsoci_sqlite3.a(vector-into-type.cpp.o): in function `soci::sqlite3_vector_into_type_backend::post_fetch(bool, soci::indicator*)':
/home/vcpkg/buildtrees/soci/src/v4.0.3-7f07d34610.clean/src/backends/sqlite3/vector-into-type.cpp:260: undefined reference to `soci::details::parse_std_tm(char const*, tm&)'
/usr/bin/ld: vcpkg/installed/x64-linux/debug/lib/libsoci_sqlite3.a(standard-into-type.cpp.o): in function `soci::sqlite3_standard_into_type_backend::post_fetch(bool, bool, soci::indicator*)':
/home/vcpkg/buildtrees/soci/src/v4.0.3-7f07d34610.clean/src/backends/sqlite3/standard-into-type.cpp:139: undefined reference to `soci::details::parse_std_tm(char const*, tm&)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/Learning_VCPKG.dir/build.make:102: Learning_VCPKG] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/Learning_VCPKG.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

### Additional context

When I asked inside discord channel somebody provided me information:

> Usually it's patched like this: https://github.com/microsoft/vcpkg/blob/78ba9711d30c64a6b40462c72f356c681e2255f3/ports/sqlpp11/fix_link_sqlite3.patch
> But soci's build system is cursed IMHO.
> Places which you would probably need to touch:
> https://github.com/SOCI/soci/blob/master/cmake/resources/SOCIConfig.cmake.in
> https://github.com/SOCI/soci/blob/master/src/backends/sqlite3/CMakeLists.txt#L13
> https://github.com/SOCI/soci/blob/master/cmake/SociBackend.cmake#L82

### PR DESCRIPTION
**Pull request by dg0yt**
Fixes #34168.
Major revision of dependency control.
The "empty" backend is now optional. 
i.e. neither `core` nor default features include any backend. (`sqlite3` might be useful. `empty` isn't.)
The `odbc` backend now has a `unixodbc` dependency when not on windows.
Adds a test-port that actually links to some backends.

Tested in x64-linux with (preview of) `vcpkg test-features soci`.

CC @baziorek @JonLiu1993

### CODE REVIEW

File: ports/soci/usage-requirements.diff [hunk 1]
```diff
@@ -1,36 +1,33 @@
...
  
          if(WIN32)
            set_target_properties(${THIS_BACKEND_TARGET}
-@@ -197,9 +203,15 @@ macro(soci_backend NAME)
+@@ -197,8 +202,14 @@ macro(soci_backend NAME)
  
          # Still need to link the libraries for tests to work
          target_link_libraries (${THIS_BACKEND_TARGET_STATIC}
 +          PUBLIC
++          ${SOCI_CORE_TARGET}_static
```
Comment (dg0yt) [inline ports/soci/usage-requirements.diff:24]: This line is the key fix for #34168.

**Comment by baziorek** [comment]
@dg0yt I've checked with provided commands (https://stackoverflow.com/questions/77140854/how-to-install-soci-db-library-with-vcpkg-on-linux) inside docker container and it is working. APPROVE:)

File: scripts/test_ports/vcpkg-ci-soci/project/main.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+#include <soci/mysql/soci-mysql.h>
+#include <soci/postgresql/soci-postgresql.h>
+#include <soci/sqlite3/soci-sqlite3.h>
+
+int main()
+{
+    soci::session mysql_db(soci::mysql, "test:mysql");
+    soci::session pgsql_db(soci::postgresql, "test:postgresql");
+    soci::session sqlite3_db(soci::sqlite3, "test.db");
+}
```
Comment (baziorek) [inline scripts/test_ports/vcpkg-ci-soci/project/main.cpp:11]: According to documentation it is supporting more backends:
image
Shouldn't we support at least all open source backends?

Comment (dg0yt) [reply scripts/test_ports/vcpkg-ci-soci/project/main.cpp:11]: Feel free to add this in another PR. You will have to package the dependencies for each backend.

Comment (dg0yt) [reply scripts/test_ports/vcpkg-ci-soci/project/main.cpp:11]: Related: https://github.com/microsoft/vcpkg/issues/15919.

**Review by JonLiu1993 — DISMISSED** [review]

**Comment by JonLiu1993** [comment]
@dg0yt, Thanks for your PR, when I tested the usage I get this error:
````
1> CMake generation started for configuration: 'x64-Debug'.
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 17 2022" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\v-zhli17\source\repos\usageTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/soci/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\v-zhli17\source\repos\usageTest" 2>&1"
1> Working directory: C:\Users\v-zhli17\source\repos\usageTest\out\build\x64-Debug
1> [CMake] -- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
1> [CMake] -- The C compiler identification is MSVC 19.36.32538.0
1> [CMake] -- The CXX compiler identification is MSVC 19.36.32538.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.36.32532/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] -- Found ZLIB: optimized;F:/Feature-test/soci/vcpkg/installed/x64-windows/lib/zlib.lib;debug;F:/Feature-test/soci/vcpkg/installed/x64-windows/debug/lib/zlibd.lib (found version "1.3.0")  
1> [CMake] -- Found OpenSSL: optimized;F:/Feature-test/soci/vcpkg/installed/x64-windows/lib/libcrypto.lib;debug;F:/Feature-test/soci/vcpkg/installed/x64-windows/debug/lib/libcrypto.lib (found version "3.1.3")  
1> [CMake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
1> [CMake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
1> [CMake] -- Looking for pthread_create in pthreads
1> [CMake] -- Looking for pthread_create in pthreads - not found
1> [CMake] -- Looking for pthread_create in pthread
1> [CMake] -- Looking for pthread_create in pthread - not found
1> [CMake] -- Found Threads: TRUE  
1> [CMake] -- Configuring done (10.3s)
1> [CMake] CMake Error at usageTest/CMakeLists.txt:9 (target_link_libraries):
1> [CMake]   Target "test" links to:
1> [CMake] 
1> [CMake]     SOCI::soci_core_static
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] 
1> [CMake] 
1> [CMake] -- Generating done (0.1s)
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 17 2022" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\v-zhli17\source\repos\usageTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/soci/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\v-zhli17\source\repos\usageTest" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 17 2022" -A x64  -DCMAKE_CONFIGURATION_TYPES:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\v-zhli17\source\repos\usageTest\out\install\x64-Debug" -DCMAKE_TOOLCHAIN_FILE=F:/Feature-test/soci/vcpkg/scripts/buildsystems/vcpkg.cmake "C:\Users\v-zhli17\source\repos\usageTest" 2>&1"' returned with exit code: 1'.

````
Please take a look.

**Comment by JonLiu1993** [comment]
All features are tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

**Comment by JonLiu1993** [comment]
Tested usage successfully by `soci[*]:x64-windows-static`:
````
soci provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(SOCI CONFIG REQUIRED)
  # note: 2 additional targets are not displayed.
  target_link_libraries(main PRIVATE SOCI::soci_core_static SOCI::soci_odbc_static SOCI::soci_empty_static SOCI::soci_mysql_static)
````

**Comment by JonLiu1993** [comment]
PR tested by user @baziorek https://github.com/microsoft/vcpkg/issues/34168#issuecomment-1752147650

**Comment by dg0yt** [comment]
It is tested by a test port...

**Comment by BillyONeal** [comment]
Amazing work as usual, thanks @dg0yt!

**Comment by baziorek** [comment]
Thanks for fixing its!
Author of corrections has priority (if he wants) to answer the question and earn some reputation:
https://stackoverflow.com/questions/77140854/how-to-install-soci-db-library-with-vcpkg-on-linux
if @dg0yt does not want - I'll answer the question:).

**Comment by dg0yt** [comment]
@baziorek I'm not active on SO.

**Comment by baziorek** [comment]
> @baziorek I'm not active on SO.

OK, I'll reply on SO.

**Comment by baziorek** [comment]
@dg0yt Something got wrong after my approve, because when I checked that time - everything was working, but when I check after merging into master (current top commit: 51475feaccb720ab40e5a0e3f8da70a4510defc0):  There are strange suggestions made by vcpkg in newest version. When I use commands (inside docker container Ubuntu22:04):
```bash
./vcpkg/vcpkg install 'soci[sqlite3]'
```
the output is:
```
soci provides CMake targets:

    find_package(soci CONFIG REQUIRED)
    target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)

    # Linking specific backends (enabled: sqlite3)
    target_link_libraries(main PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_<BACKEND>,SOCI::soci_<BACKEND>,SOCI::soci_<BACKEND>_static>)
```
when I use those commands in CMake:
```
./vcpkg/vcpkg integrate install
cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake && make
```
I see:
```
Soci found:  and 
-- Configuring done
-- Generating done
-- Build files have been written to: /home
[ 50%] Building CXX object CMakeFiles/Learning_VCPKG.dir/main.cpp.o
[100%] Linking CXX executable Learning_VCPKG
/usr/bin/ld: CMakeFiles/Learning_VCPKG.dir/main.cpp.o: warning: relocation against `_ZN4soci7sqlite3E' in read-only section `.text'
/usr/bin/ld: CMakeFiles/Learning_VCPKG.dir/main.cpp.o: in function `main':
main.cpp:(.text+0x61): undefined reference to `soci::sqlite3'
/usr/bin/ld: warning: creating DT_TEXTREL in a PIE
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/Learning_VCPKG.dir/build.make:99: Learning_VCPKG] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/Learning_VCPKG.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

CMakeLists.txt (with suggestion used):
```
cmake_minimum_required(VERSION 3.16)
project(Learning_VCPKG LANGUAGES CXX)

find_package(SOCI CONFIG REQUIRED)

if (SOCI_FOUND)
    message("Soci found: ${SOCI_INCLUDE_DIRS} and ${SOCI_LIBRARY_DIRS}")
endif ()

add_executable(${PROJECT_NAME} main.cpp)

# find_package(soci CONFIG REQUIRED)
target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
```

and `main.cpp` one more time:
```c++
#include <exception>
#include <iostream>
#include <soci/soci.h>
#include <soci/sqlite3/soci-sqlite3.h>
#include <string>

using namespace soci;
using namespace std;

int main()
{
    try {
        soci::session sql(sqlite3, "test.db");

        sql << "CREATE TABLE IF NOT EXISTS Employees("
               "ID     INT PRIMARY KEY NOT NULL,"
               "name   TEXT            NOT NULL,"
               "Age   INT             NOT NULL,"
               "Address  TEXT,"
               "Salary REAL);";
        sql << "INSERT INTO Employees (ID, name, Age, Address, Salary) VALUES (1, 'Tadeusz', 24, 'Krakow', 2000.01 ); ";
        sql << "INSERT INTO Employees (ID, name, Age, Address, Salary) VALUES (2, 'Symeon', 25, 'Texas', 15000.00 ); ";
        sql << "INSERT INTO Employees (ID, name, Age, Address, Salary) VALUES (3, 'Rodion', 23, 'Norway', 20000.00 );";

        int count;
        sql << "select count(*) from Employees", into(count);

        cout << "We have " << count << " entries in the Employees table.\n";
    } catch (exception const& e) {
        cerr << "Error: " << e.what() << '\n';
    }
}
```

**Comment by dg0yt** [comment]
> ~~~
> target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
> ~~~

You are linking core only. In a static build, you have to link the backend you want.

**Comment by baziorek** [comment]
> > ```
> > target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
> > ```
> 
> You are linking core only. ~In a static build~ When directly accessing a particular backend, you have to link that backend.

When I change the CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.16)
project(Learning_VCPKG LANGUAGES CXX)

find_package(SOCI CONFIG REQUIRED)

if (SOCI_FOUND)
    message("Soci found: ${SOCI_INCLUDE_DIRS} and ${SOCI_LIBRARY_DIRS}")
endif ()

add_executable(${PROJECT_NAME} main.cpp)

target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_<BACKEND>,SOCI::soci_<BACKEND>,SOCI::soci_<BACKEND>_static>)
```
I see compilation problem:
```
cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake && make
Soci found:  and 
-- Configuring done
CMake Error at CMakeLists.txt:14 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_<BACKEND>

  $<TARGET_EXISTS:tgt> expression requires a non-empty valid target name.


CMake Error at CMakeLists.txt:14 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_<BACKEND>

  $<TARGET_EXISTS:tgt> expression requires a non-empty valid target name.


CMake Error at CMakeLists.txt:14 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_<BACKEND>

  $<TARGET_EXISTS:tgt> expression requires a non-empty valid target name.


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
___________________
So I tried to replace `<BACKEND>` with `sqlite3`, so CMake is now:
```
cmake_minimum_required(VERSION 3.16)
project(Learning_VCPKG LANGUAGES CXX)

find_package(SOCI CONFIG REQUIRED)

if (SOCI_FOUND)
    message("Soci found: ${SOCI_INCLUDE_DIRS} and ${SOCI_LIBRARY_DIRS}")
endif ()

add_executable(${PROJECT_NAME} main.cpp)

target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
```
And compilation also failures:
```
cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake && make
Soci found:  and 
-- Configuring done
CMake Error at CMakeLists.txt:13 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>

  $<TARGET_EXISTS> expression requires exactly one parameter.


CMake Error at CMakeLists.txt:13 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>

  $<TARGET_EXISTS> expression requires exactly one parameter.


CMake Error at CMakeLists.txt:13 (target_link_libraries):
  Error evaluating generator expression:

    $<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>

  $<TARGET_EXISTS> expression requires exactly one parameter.


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Comment by dg0yt** [comment]
~~~diff
-target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
+target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3>,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
~~~

**Comment by baziorek** [comment]
> ```diff
> -target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
> +target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3>,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
> ```

Thanks, when I correct as You suggested CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.16)
project(Learning_VCPKG LANGUAGES CXX)

find_package(SOCI CONFIG REQUIRED)

if (SOCI_FOUND)
    message("Soci found: ${SOCI_INCLUDE_DIRS} and ${SOCI_LIBRARY_DIRS}")
endif ()

add_executable(${PROJECT_NAME} main.cpp)

target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_core>,SOCI::soci_core,SOCI::soci_core_static>)
target_link_libraries(${PROJECT_NAME} PRIVATE $<IF:$<TARGET_EXISTS:SOCI::soci_sqlite3>,SOCI::soci_sqlite3,SOCI::soci_sqlite3_static>)
```
everything works:
```
cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake && make
Soci found:  and 
-- Configuring done
-- Generating done
-- Build files have been written to: /home
Consolidate compiler generated dependencies of target Learning_VCPKG
[ 50%] Linking CXX executable Learning_VCPKG
[100%] Built target Learning_VCPKG
root@agh-legion:/home# cmake . -DCMAKE_TOOLCHAIN_FILE=/home/vcpkg/scripts/buildsystems/vcpkg.cmake && make && ./Learning_VCPKG 
Soci found:  and 
-- Configuring done
-- Generating done
-- Build files have been written to: /home
Consolidate compiler generated dependencies of target Learning_VCPKG
[100%] Built target Learning_VCPKG
We have 3 entries in the Employees table.
```
______________
But IMO suggestion from VCPKG needs modification, when I compare diff between suggestion and working version:
image
First occurrence of text `<BACKEND` should be replaced with `sqlite3`, but another two occurrences to replace are: `<BACKEND>` (extra `>`). IMO it should be the same text to replace in all three places.
Of course it would be much easier if suggestion will be just copy&paste without replacing `<BACKEND>`  keyword, but maybe it is impossible:)

**Comment by dg0yt** [comment]
I see, we lost a `>`. I want to avoid `<...>` for BACKEND, but I couldn't come up with an alternative fast enough. I avoided adding `usage` in the beginning because it wouldn't be trivial, but the uninformed heuristics is helpful either.

**Pull request merged**